projectq/cengines/_replacer/_decompositionhandling.py

```diff
@@ -11,67 +11,101 @@
 #   limitations under the License.
 
 from ._decomposition import Decomposition
+from projectq.ops import BasicGate
 
-decompositions = dict()
 
-
-class ThisIsNotAGateClassError(Exception):
+class ThisIsNotAGateClassError(TypeError):
     pass
 
 
-def register_decomposition(gate_class, gate_decomposer, gate_recognizer=None):
+class DecompositionRule:
```

Comment:
@Strilanc I'm sorry about the long wait, github hid this thread from me.

I still think it would be nicer if `DecompositionRuleSet` actually contained `DecompositionRule` objects (instead of `_Decomposition` objects). Why not combine them into one class?